pkg/virt-launcher/virtwrap/live-migration-source.go

```diff
@@ -176,91 +174,82 @@ func generateDomainForTargetCPUSetAndTopology(vmi *v1.VirtualMachineInstance, do
 	return domain, err
 }
 
-func injectNewSection(encoder *xml.Encoder, domain *api.Domain, section []string, logger *log.FilteredLogger) error {
-	// Marshalling the whole domain, even if we just need the cputune section, for indentation purposes
-	xmlstr, err := xml.MarshalIndent(domain.Spec, "", "  ")
-	if err != nil {
-		logger.Reason(err).Error("Live migration failed. Failed to get XML.")
-		return err
+// TODO: remove all the convert function once we switch of using the types of libvirtxml.
+// See: https://github.com/kubevirt/kubevirt/issues/10844
+func convertKubeVirtCPUFeatureToDomainCPUFeature(features []api.CPUFeature) []libvirtxml.DomainCPUFeature {
+	var res []libvirtxml.DomainCPUFeature
+	for _, f := range features {
+		res = append(res, libvirtxml.DomainCPUFeature{
+			Name:   f.Name,
+			Policy: f.Policy,
+		})
 	}
-	decoder := xml.NewDecoder(bytes.NewReader(xmlstr))
-	var location = make([]string, 0)
-	var newLocation []string = nil
-	injecting := false
-	for {
-		if newLocation != nil {
-			// Postpone popping end elements from `location` to ensure their removal
-			location = newLocation
-			newLocation = nil
-		}
-		token, err := decoder.RawToken()
-		if err == io.EOF {
-			break
-		}
-		if err != nil {
-			logger.Errorf("error getting token: %v\n", err)
-			return err
-		}
+	return res
+}
 
-		switch v := token.(type) {
-		case xml.StartElement:
-			location = append(location, v.Name.Local)
+func convertKubeVirtCPUTopologyToDomainCPUTopology(cpuTopology *api.CPUTopology) *libvirtxml.DomainCPUTopology {
+	if cpuTopology == nil {
+		return nil
+	}
+	return &libvirtxml.DomainCPUTopology{
+		Sockets: int(cpuTopology.Sockets),
+		Cores:   int(cpuTopology.Cores),
+		Threads: int(cpuTopology.Threads),
+	}
+}
 
-		case xml.EndElement:
-			newLocation = location[:len(location)-1]
+func convertKubeVirtNUMACellToDomainDomainCell(cell []api.NUMACell) []libvirtxml.DomainCell {
+	var ret []libvirtxml.DomainCell
+	for _, c := range cell {
+		v, err := strconv.ParseUint(c.ID, 10, 32)
+		if err != nil {
+			continue
```

Comment:
IMHO, better if we handle the error case properly instead of silently ignoring it. The caller should return an error if this conversion fails.

Reply:
You are right, I updated the PR